mypy/semanal.py

```diff
@@ -1674,12 +1679,32 @@ def anal_type(self, t: Type, *,
                   tvar_scope: Optional[TypeVarScope] = None,
                   allow_tuple_literal: bool = False,
                   aliasing: bool = False,
-                  third_pass: bool = False) -> Type:
+                  third_pass: bool = False,
+                  from_bases: Optional[ClassDef] = None) -> Type:
         a = self.type_analyzer(tvar_scope=tvar_scope,
                                aliasing=aliasing,
                                allow_tuple_literal=allow_tuple_literal,
                                third_pass=third_pass)
-        return t.accept(a)
+        tp = t.accept(a)
+        if a.aliases_used:
+            if from_bases is not None:
+                self.cur_mod_node.alias_deps[from_bases].update(a.aliases_used)
+                return tp
+            self.add_alias_deps(a.aliases_used)
+        return tp
+
+    def add_alias_deps(self, aliases_used: Set[str]) -> None:
+        """Add full names of type aliases on which the current node depends.
+
+        This is used by fine-grained incremental mode to re-chek the corresponding nodes.
+        """
+        if self.is_class_scope():
```

Comment:
There is no `cur_node` in pass two (only in pass three). Instead, in pass two there are two separate scope stacks: function stack and class stack. Maybe this can be simplified, but I think it would be a large refactoring for relatively little benefits.